lib/teleterm/vnet/service.go

```diff
@@ -14,35 +14,263 @@
+		return trace.BadParameter("missing DaemonService")
+	}
+
+	if c.ClientStore == nil {
+		return trace.BadParameter("missing ClientStore")
+	}
+
+	return nil
 }
 
 func (s *Service) Start(ctx context.Context, req *api.StartRequest) (*api.StartResponse, error) {
-	n := rand.Intn(10)
-	randomDelay := time.Duration(n) * 100 * time.Millisecond
-	time.Sleep(randomDelay + 400*time.Millisecond)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.status == statusClosed {
+		return nil, trace.CompareFailed("VNet service has been closed")
+	}
+
+	if s.status == statusRunning {
+		return &api.StartResponse{}, nil
+	}
+
+	appProvider := &appProvider{
+		daemonService:      s.cfg.DaemonService,
+		clientStore:        s.cfg.ClientStore,
+		insecureSkipVerify: s.cfg.InsecureSkipVerify,
+	}
+
+	processManager, err := vnet.SetupAndRun(ctx, appProvider)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	go func() {
+		err := processManager.Wait()
+		if err != nil && !errors.Is(err, context.Canceled) {
+			log.ErrorContext(ctx, "VNet closed with an error", "error", err)
+		} else {
+			log.DebugContext(ctx, "VNet closed")
+		}
+
+		// TODO(ravicious): Notify the Electron app about change of VNet state, but only if it's
+		// running. If it's not running, then the Start RPC has already failed and forwarded the error
+		// to the user.
+
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		if s.status == statusRunning {
+			s.status = statusNotRunning
+		}
+	}()
+
+	s.processManager = processManager
+	s.status = statusRunning
 	return &api.StartResponse{}, nil
 }
 
+// Stop stops VNet and cleans up used resources. Blocks until VNet stops or ctx is canceled.
 func (s *Service) Stop(ctx context.Context, req *api.StopRequest) (*api.StopResponse, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := s.stopLocked()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	return &api.StopResponse{}, nil
 }
 
-// Close stops the current VNet instance and prevents new instances from being started.
-//
+func (s *Service) stopLocked() error {
+	if s.status == statusClosed {
+		return trace.CompareFailed("VNet service has been closed")
+	}
+
+	if s.status == statusNotRunning {
+		return nil
+	}
+
+	s.processManager.Close()
+	err := s.processManager.Wait()
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return trace.Wrap(err)
+	}
+
+	s.status = statusNotRunning
+	return nil
+}
+
+// Close stops VNet service and prevents it from being started again. Blocks until VNet stops.
 // Intended for cleanup code when tsh daemon gets terminated.
 func (s *Service) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := s.stopLocked()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.status = statusClosed
 	return nil
 }
+
+type appProvider struct {
+	daemonService      *daemon.Service
+	clientStore        *client.Store
+	insecureSkipVerify bool
+}
+
+func (p *appProvider) ListProfiles() ([]string, error) {
+	profiles, err := p.clientStore.ListProfiles()
```

Comment:
would it be worth getting rid of clientStore and using something like `p.daemonService.ListRootClusters`? I guess that's a bit heavier since it reads all the profiles - it just feels a bit off to circumvent the storage layer you guys have and reaching down to the filesystem here. We could probably add a lighter-weight method to list the profiles in daemonService, or possibly pass the storage layer right in here

Reply:
Yeah, that makes sense. I just removed `client.Store` and replaced calls to it with `daemon.Service`, I just had to add that one method that merely lists the profile names.

Initially I like that we can just pass `client.Store`, which feels much more light than this entangled mess that `daemon.Service` is. But using `daemon.Service` instead does help to remove some extra methods from `appProvider` which were there only because we were using `client.Store`, such as `newTeleportClient`.

Even more important, we wouldn't actually want people to use the client from `newTeleportClient` to make calls. The one returned by `daemon.Service` is wired up to MFA handlers and so on. This is also the reason why passing the storage layer would not be a great idea.